# <center>Đồ án 3: Linear Regression</center>

# Thông tin sinh viên

- Họ và tên: Bùi Đỗ Duy Quân
- MSSV: 21127141
- Lớp: 21CLC02

# Import

In [52]:
import pandas as pd
import numpy as np

# Import thêm dữ thư viện nếu cần

# Đọc dữ liệu

In [53]:
# Đọc dữ liệu bằng pandas
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

# Lấy các đặc trưng X và giá trị mục tiêu y cho các tập huấn luyện (train) và kiểm tra (test)
X_train = train.iloc[:, :-1]    # Dataframe (chứa các đặc trưng huấn luyện)
y_train = train.iloc[:, -1]     # Series    (chứa 1 giá trị mục tiêu kiểm tra)

X_test = test.iloc[:, :-1]      # Dataframe (chứa các đặc trưng kiểm tra)
y_test = test.iloc[:, -1]       # Series    (chứa 1 giá trị mục tiêu kiểm tra)

# Sinh viên có thể sử dụng các khác nếu cần

# Cài đặt hàm

In [54]:
# Cài đặt các hàm cần thiết ở đây
class OLSLinearRegression:
    def fit(self, X, y):
        #X_pinv is the pseudoinverse of X, which is defined as X_pinv = (X.T * X)^(-1) * X.T
        X_pinv = np.linalg.inv(X.T @ X) @ X.T    # np.linalg.pinv(X)
        self.w = X_pinv @ y

        return self

    
    def get_params(self):
        return self.w


    def predict(self, X):
        # X is a matrix of shape (n_samples, n_features)
        # w is a vector of shape (n_features, )
        # the result is: y = X * w, a vector of shape (n_samples, )
        #.ravel() is used to convert a matrix into a vector
        # vector multiplication with matrix is the same as matrix multiplication with matrix
        return np.sum(self.w.ravel() * X, axis=1)   

In [55]:
def mae(y, y_hat):
    return np.mean(np.abs(y.ravel() - y_hat.ravel()))

# Yêu cầu 1a: Sử dụng toàn bộ 11 đặc trưng đầu tiên `Gender`, `10percentage`, `12percentage`, `CollegeTier`, `Degree`, `collegeGPA`, `CollegeCityTier`, `English`, `Logical`, `Quant`, `Domain` (2 điểm) 

In [56]:
# Phần code cho yêu cầu 1a
new_features = X_train.iloc[:, :11]
x = new_features.to_numpy()
y = y_train.to_numpy().reshape(-1, 1)



lr = OLSLinearRegression().fit(x, y)
lr.get_params()
y_pre = lr.predict(X_test.iloc[:, :11].to_numpy())

In [57]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra
print("MAE: ", mae(y_test, y_pre))

MAE:  104863.77754033315


Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = (-22756.513)*X_1 + 804.503*X_2+ 1294.655*X_3 + (-91781.898)*X_4 + 23182.389*X_5  + 1437.549*X_6 + (-8570.662)*X_7 + 147.858*X_8 + \\152.888*X_9 + 117.222*X_{10} + 34552.286*X_{11}$$

# Yêu cầu 1b: Xây dựng mô hình sử dụng duy nhất 1 đặc trưng tính cách với các đặc trưng tính cách gồm `conscientiousness`, `agreeableness`, `extraversion`, `nueroticism`, `openess_to_experience`, tìm mô hình cho kết quả tốt nhất (1 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ đặc trưng

In [58]:
def getTrain(index, folks):
    Train = np.empty((0, folks[0].shape[1]))
    for i in range(len(folks)):
        if i != index:
            Train = np.concatenate((Train, folks[i]))
    return Train

In [59]:
# Phần code cho yêu cầu 1b
# Tìm ra đặc trưng tốt nhất
# In ra các kết quả cross-validation như yêu cầu

def Best_Feature_Personality(Df_np, k_cluster):
    folks = np.array_split(Df_np, k_cluster) # split data to 10 folds
    mae_arr = np.zeros((k_cluster, 5)) # array to store mae of each fold
    for i in range(k_cluster):
        data_train = getTrain(i, folks) # get train data (9 folds except i-th fold)
        data_test = folks[i]            # get test data (i-th fold)

        X_test, y_test = data_test[:, :-1], data_test[:, -1]  #all features and salary in test data

        X_train, y_train = data_train[:, :-1], data_train[:, -1] #all features in each folds and salary in train data


        for j in range(-5, 0):
            feature_train, feature_test = X_train[:, j], X_test[:, j]    # 1 of 5 features in train and test data

            lr = OLSLinearRegression().fit(feature_train[:, None], y_train[:, None]) # fit feature to salary
            
            y_pre = lr.predict(feature_test[:, None]) # predict salary
            mae_feature = mae(y_test[:, None], y_pre[:, None]) # calculate mae
            mae_arr[i, j+5] = mae_feature

    result = np.mean(mae_arr, axis=0) # mean of mae in each feature
    print("MAE of each feature: ", result)
    return np.argmin(result) # return index of best feature

In [60]:
# Huấn luyện lại mô hình best_personality_feature_model với đặc trưng tốt nhất trên toàn bộ tập huấn luyện
def best_personality_feature_model():
    personality = ["conscientiousness", "agreeableness", "extraversion", "nueroticism", "openess_to_experience"]
    Df_np = train.to_numpy()
    np.random.default_rng().shuffle(Df_np, axis=0) # shuffle data

    feature = personality[Best_Feature_Personality(Df_np, 10)]

    print("Best feature: ", feature)
    X_train = train.loc[:, feature].to_numpy().reshape(-1, 1)

    X_test = test.loc[:, feature].to_numpy().reshape(-1, 1)

    lr = OLSLinearRegression().fit(X_train, y_train)
    return lr.predict(X_test), lr.get_params()

In [61]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình best_personality_feature_model
y_pre, params = best_personality_feature_model()
print("MAE: ", mae(y_test, y_pre))
print("Params: ", params)

MAE of each feature:  [306226.69154922 300790.04869829 306998.20881964 299281.34116466
 302984.14166465]
Best feature:  nueroticism
MAE:  291019.693226953
Params:  [-56546.304]


Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = (-56546.304)*neuroticism$$

# Yêu cầu 1c: Xây dựng mô hình sử dụng duy nhất 1 đặc trưng `English`, `Logical`, `Quant`, tìm mô hình cho kết quả tốt nhất (1 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ đặc trưng

In [62]:
# Phần code cho yêu cầu 1c
# Tìm ra đặc trưng tốt nhất
# In ra các kết quả cross-validation như yêu cầu
Feature_index = [train.columns.get_loc(c) for c in train.columns if c in ["English", "Logical", "Quant"]]

def Best_Feature_Skill(Df_np, k_cluster):
    folks = np.array_split(Df_np, k_cluster) # split data to 10 folds
    mae_arr = np.zeros((k_cluster, 3)) # array to store mae of each fold

    for i in range(k_cluster):
        data_train = getTrain(i, folks) # get train data (9 folds except i-th fold)
        data_test = folks[i]            # get test data (i-th fold)

        X_test, y_test = data_test[:, :-1], data_test[:, -1]  #all features and salary in test data

        X_train, y_train = data_train[:, :-1], data_train[:, -1] #all features in each folds and salary in train data

        col = 0
        for j in Feature_index:
            feature_train, feature_test = X_train[:, j], X_test[:, j]    # 1 of 5 features in train and test data

            lr = OLSLinearRegression().fit(feature_train[:, None], y_train[:, None]) # fit feature to salary
            
            y_pre = lr.predict(feature_test[:, None]) # predict salary
            mae_feature = mae(y_test[:, None], y_pre[:, None]) # calculate mae
            mae_arr[i, col] = mae_feature
            col += 1

    result = np.mean(mae_arr, axis=0) # mean of mae in each feature
    print("MAE of each feature: ", result.round(3))
    return np.argmin(result) # return index of best feature

In [63]:
# Huấn luyện lại mô hình best_skill_feature_model với đặc trưng tốt nhất trên toàn bộ tập huấn luyện
def best_skill_feature_model():
    Df_np = train.to_numpy()
    np.random.default_rng().shuffle(Df_np, axis=0) # shuffle data

    feature = Feature_index[Best_Feature_Skill(Df_np, 10)]

    print("Best feature: ", train.columns[feature])
    X_train = train.iloc[:, feature].to_numpy().reshape(-1, 1)

    X_test = test.iloc[:, feature].to_numpy().reshape(-1, 1)

    lr = OLSLinearRegression().fit(X_train, y_train)
    return lr.predict(X_test), lr.get_params()

In [64]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình best_skill_feature_model
y_pre, params = best_skill_feature_model()
print("MAE: ", mae(y_test, y_pre))
print("Params: ", params)

MAE of each feature:  [121982.572 120312.947 118133.097]
Best feature:  Quant
MAE:  106819.5776198967
Params:  [585.895]


Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = (585.895) * Quant$$

# Yêu cầu 1d: Sinh viên tự xây dựng mô hình, tìm mô hình cho kết quả tốt nhất (3 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ $m$ mô hình mà sinh viên thiết kế

## Tìm mô hình

In [65]:
# Trình bày các phần tìm ra mô hình

## Thử nghiệm, so sánh các mô hình

In [66]:
# Phần code cho yêu cầu 1d
# Tìm ra mô hình tốt nhất (tự thiết kế bởi sinh viên)
# In ra các kết quả cross-validation như yêu cầu

In [67]:
# Huấn luyện lại mô hình my_best_model trên toàn bộ tập huấn luyện

In [68]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình my_best_model

Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = ...$$